CMS-Enterprise/sbom-harbor-ui
# test: add global mocks, cover main.tsx

# Pull Request 

## Summary

This PR introduces significant test coverage for the `main.tsx` module and enhances our testing setup by importing global mocks for `aws-amplify` and `web-vitals` in Jest. This approach eliminates the redundancy of duplicating `aws-amplify` mocks across multiple test files. 

**Key Changes:**

1. The `onPerfEntry` function, previously encapsulated within an IIFE in `main`, has been extracted to its own separate module. This modular approach promotes better code organization and readability.
2. The named export `CONFIG` from `src/utils/config.ts` has been converted to a default export. This change has been propagated to all files where it is imported, ensuring that the application continues to function as expected.
3. The `main` module has been updated to conditionally log web-vitals to the console only when `NODE_ENV === 'development'`. This change prevents unnecessary console logging in production environments, contributing to overall performance and user experience.
4. The `--all` flag has been removed from the `yarn test` script. As a result, running `yarn test -- path/to/some/test.tsx` will now only output coverage for the specific files passed to the script after the `--`, instead of all files. This tweak makes it easier for developers to focus on specific tests without being distracted by the coverage of the entire codebase.

### Added

- `main.test.tsx`
- `utils/onPerfEntry.ts`
- `utils/onPerfEntry.test.ts`

## Changed

1. Extracted `onPerfEntry` function from `main` to a separate module.
2. Changed the named export `CONFIG` from `src/utils/config.ts` to a default export and updated all import references.
3. Updated `main` to log web-vitals to the console only when `NODE_ENV === 'development'`.
4. Removed `--all` flag from `yarn test` script.

### Deprecated

- n/a

### Removed

- n/a

### Fixed

- n/a

## How to test

To test these changes, you can use the following commands:

1. Run `yarn lint` to check for any linting errors.
2. Run `yarn test` to execute the test suite.
3. Run `yarn build` to ensure that the application builds correctly.
4. Run `yarn start` and log in to ensure that the application is mounting correctly.

**Test Coverage:**

```
-------------------------------------|---------|----------|---------|---------|---------------------------
File                                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-------------------------------------|---------|----------|---------|---------|---------------------------
All files                            |   57.96 |    24.91 |    19.2 |   57.67 |
 src                                 |   98.63 |      100 |     100 |     100 |
  App.tsx                            |     100 |      100 |     100 |     100 |
  constants.ts                       |    87.5 |      100 |     100 |     100 |
  main.tsx                           |     100 |      100 |     100 |     100 |
  types.ts                           |     100 |      100 |     100 |     100 |
 src/actions                         |     100 |      100 |     100 |     100 |
  actionTypes.ts                     |     100 |      100 |     100 |     100 |
  loginUser.tsx                      |     100 |      100 |     100 |     100 |
  loginUserFederated.tsx             |     100 |      100 |     100 |     100 |
  logoutUser.tsx                     |     100 |      100 |     100 |     100 |
 src/utils                           |   80.21 |     67.5 |      75 |   80.24 |
  config.ts                          |     100 |      100 |     100 |     100 |
  configureCognito.ts                |   88.88 |      100 |     100 |    87.5 | 9
  onPerfEntry.ts                     |     100 |      100 |     100 |     100 |
  ```
